external/fv3fit/fv3fit/pytorch/cyclegan/cyclegan_trainer.py

```diff
@@ -0,0 +1,424 @@
+from fv3fit.pytorch.system import DEVICE
+import itertools
+import numpy as np
+
+
+@dataclasses.dataclass
+class CycleGANNetworkConfig:
+    """
+    Configuration for building and training a CycleGAN network.
+
+    Attributes:
+        generator_optimizer: configuration for the optimizer used to train the
+            generator
+        discriminator_optimizer: configuration for the optimizer used to train the
+            discriminator
+        generator: configuration for building the generator network
+        discriminator: configuration for building the discriminator network
+        identity_loss: loss function used to make the generator which outputs
+            a given domain behave as an identity function when given data from
+            that domain as input
+        cycle_loss: loss function used on the difference between a round-trip
+            of the CycleGAN network and the original input
+        gan_loss: loss function used on output of the discriminator when
+            training the discriminator identify samples correctly or when training
+            the generator to fool the discriminator
+        identity_weight: weight of the identity loss
+        cycle_weight: weight of the cycle loss
+        generator_weight: weight of the generator's gan loss
+        discriminator_weight: weight of the discriminator gan loss
+    """
+
+    generator_optimizer: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    discriminator_optimizer: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("Adam")
+    )
+    generator: "GeneratorConfig" = dataclasses.field(
+        default_factory=lambda: GeneratorConfig()
+    )
+    discriminator: "DiscriminatorConfig" = dataclasses.field(
+        default_factory=lambda: DiscriminatorConfig()
+    )
+    identity_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    cycle_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    gan_loss: LossConfig = dataclasses.field(default_factory=LossConfig)
+    identity_weight: float = 0.5
+    cycle_weight: float = 1.0
+    generator_weight: float = 1.0
+    discriminator_weight: float = 1.0
+
+    def build(
+        self, n_state: int, n_batch: int, state_variables, scalers
+    ) -> "CycleGANTrainer":
+        generator_a_to_b = self.generator.build(n_state)
+        generator_b_to_a = self.generator.build(n_state)
+        discriminator_a = self.discriminator.build(n_state)
+        discriminator_b = self.discriminator.build(n_state)
+        optimizer_generator = self.generator_optimizer.instance(
+            itertools.chain(
+                generator_a_to_b.parameters(), generator_b_to_a.parameters()
+            )
+        )
+        optimizer_discriminator = self.discriminator_optimizer.instance(
+            itertools.chain(discriminator_a.parameters(), discriminator_b.parameters())
+        )
+        return CycleGANTrainer(
```

Comment:
CycleGANTrainer is "module public" because it's used in `train_cyclegan.py`, but it's package-private because it's not imported to `fv3fit.pytorch`. I can rename it to `Trainer`, it's just that we have a lot of different objects named things like "train" (e.g. multiple TrainingLoopConfig), and this object's motivation for existing is that cyclegan training is complicated. I'm not sure which name is better.

CycleGAN (and by extension CycleGANModule) are package-public because I will be importing them as symbols in code that evaluates CycleGAN output (which will definitely need to be bespoke compared to evaluating other models).

Reply:
OK that's probably fine, I agree that it's not obvious that dropping the prefix makes the names all that much clearer.